SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/ReferenceCountedLockingPolicy.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "ReferenceCountedLockingPolicy.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "Interop/SpatialStaticComponentView.h"
+#include "Schema/AuthorityIntent.h"
+
+#include "GameFramework/Actor.h"
+
+DEFINE_LOG_CATEGORY(LogReferenceCountedLockingPolicy);
+
+ActorLockToken UReferenceCountedLockingPolicy::AcquireLock(const AActor* Actor, FString DebugString)
+{
+	check(Actor != nullptr);
+	if (!CanAcquireLock(Actor)) {
+		UE_LOG(LogReferenceCountedLockingPolicy, Error, TEXT("Called AcquireLock when CanAcquireLock returned false. Actor: %s."), *Actor->GetName());
+		return SpatialConstants::INVALID_ACTOR_LOCK_TOKEN;
+	}
+	++ActorToReferenceCount.FindOrAdd(Actor);
+	UE_LOG(LogReferenceCountedLockingPolicy, Log, TEXT("Acquiring migration lock. "
+		"Actor: %s. Lock name: %s. Token %d: Locks held: %d."), *Actor->GetName(), *DebugString, NextToken, *ActorToReferenceCount.Find(Actor));
+	TokenToNameAndActor.Emplace(NextToken, LockNameAndActor{ MoveTemp(DebugString), Actor });
+	return NextToken++;
+}
+
+bool UReferenceCountedLockingPolicy::CanAcquireLock(const AActor* Actor) const
+{
+	check(Actor != nullptr);
+	const auto* NetDriver = Cast<USpatialNetDriver>(Actor->GetWorld()->GetNetDriver());
+	const auto EntityId = NetDriver->PackageMap->GetEntityIdFromObject(Actor);
+
+	const bool HasAuthIntent = NetDriver->VirtualWorkerTranslator->GetLocalVirtualWorkerId() ==
+			NetDriver->StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(EntityId)->VirtualWorkerId;
```

Comment:
Are we always guaranteed to get a AuthorityIntent component here? What if the entity id is invalid? Or the component hasn't been checked out yet?

Reply:
Ah, hadn't considered trying to lock an entity that doesn't exist yet. Maybe it should also check if it's in the static component view.
It might be worth remembering this as a concrete difference between unreal and spatial authority and we might need to lean into that more.

Reply:
I can't imagine a situation in which the AuthorityIntent doesn't exist on an entity we're trying to check but entity id being invalid definitely makes sense, will add a HasComponent check also

Reply:
That is actually fairly upsetting as it is a concrete case where we can't just pretend we have an entity.
If we guarantee that the same worker will always eventually get authority we can deal with this by never losing it. But this is a case where locking doesn't do the thing the user might think it does.